File: go/nemo_relay/pricing.go
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package nemo_relay

import "encoding/json"

// PricingPluginKind is the top-level plugin kind used by the pricing component.
const PricingPluginKind = "pricing"

// Pricing unit constants accepted by pricing catalog entries.
const (
	PricingUnitPerToken   = "per_token"
	PricingUnitPerRequest = "per_request"
	PricingUnitPerSecond  = "per_second"
	PricingUnitGPUHour    = "gpu_hour"
)

// Prompt-cache accounting constants accepted by pricing catalog entries.
const (
	CacheReadAccountingIncludedInPromptTokens = "included_in_prompt_tokens"
	CacheReadAccountingSeparate               = "separate"
)

// PricingConfig is the canonical Go shape for the pricing plugin config document.
type PricingConfig struct {
	Sources []PricingSourceConfig `json:"sources,omitempty"`
}

// PricingSourceConfig is implemented by pricing source config structs.
type PricingSourceConfig interface {
	pricingSourceConfig()
}

// PricingInlineSourceConfig embeds a pricing catalog directly in plugin config.
type PricingInlineSourceConfig struct {
	Catalog PricingCatalog `json:"catalog"`
}

func (PricingInlineSourceConfig) pricingSourceConfig() {}

// MarshalJSON serializes the inline source with the canonical type discriminator.
func (source PricingInlineSourceConfig) MarshalJSON() ([]byte, error) {
	type alias PricingInlineSourceConfig
	return json.Marshal(struct {
		Type string `json:"type"`
		alias
	}{
		Type:  "inline",
		alias: alias(source),
	})
}

// PricingFileSourceConfig loads a pricing catalog from a JSON file path.
type PricingFileSourceConfig struct {
	Path string `json:"path"`
}

func (PricingFileSourceConfig) pricingSourceConfig() {}

// MarshalJSON serializes the file source with the canonical type discriminator.
func (source PricingFileSourceConfig) MarshalJSON() ([]byte, error) {
	type alias PricingFileSourceConfig
	return json.Marshal(struct {
		Type string `json:"type"`
		alias
	}{
		Type:  "file",
		alias: alias(source),
	})
}

// PricingCatalog is an inline pricing catalog payload.
type PricingCatalog struct {
	Version uint32         `json:"version,omitempty"`
	Entries []ModelPricing `json:"entries,omitempty"`
}

// ModelPricing is one model pricing catalog entry.
type ModelPricing struct {
	Provider      string                            `json:"provider"`
	ModelID       string                            `json:"model_id"`
	Aliases       []string                          `json:"aliases,omitempty"`
	Currency      string                            `json:"currency,omitempty"`
	Unit          string                            `json:"unit,omitempty"`
	Rates         *TokenPricingRates                `json:"rates,omitempty"`
	RateSchedule  *PromptTokenThresholdRateSchedule `json:"rate_schedule,omitempty"`
	PromptCache   PromptCachePricing                `json:"prompt_cache"`
	PricingAsOf   string                            `json:"pricing_as_of"`
	PricingSource string                            `json:"pricing_source"`
}

// TokenPricingRates expresses model rates per one million tokens.
type TokenPricingRates struct {
	InputPerMillion      float64  `json:"input_per_million"`
	OutputPerMillion     float64  `json:"output_per_million"`
	CacheReadPerMillion  *float64 `json:"cache_read_per_million,omitempty"`
	CacheWritePerMillion *float64 `json:"cache_write_per_million,omitempty"`
}

// PromptCachePricing configures cache-read token accounting for a pricing entry.
type PromptCachePricing struct {
	ReadAccounting string `json:"read_accounting"`
}

// TokenRateTier is one prompt-token threshold tier.
type TokenRateTier struct {
	MinPromptTokens *uint64           `json:"min_prompt_tokens,omitempty"`
	MaxPromptTokens *uint64           `json:"max_prompt_tokens,omitempty"`
	Rates           TokenPricingRates `json:"rates"`
}

// PromptTokenThresholdRateSchedule selects token rates by full-request prompt token thresholds.
type PromptTokenThresholdRateSchedule struct {
	AppliesTo string          `json:"applies_to,omitempty"`
	Tiers     []TokenRateTier `json:"tiers"`
}

// MarshalJSON serializes the rate schedule with the canonical type discriminator.
func (schedule PromptTokenThresholdRateSchedule) MarshalJSON() ([]byte, error) {
	type alias PromptTokenThresholdRateSchedule
	return json.Marshal(struct {
		Type string `json:"type"`
		alias
	}{
		Type:  "prompt_token_threshold",
		alias: alias(schedule),
	})
}

// PricingComponentSpec wraps one pricing config as a top-level plugin component.
type PricingComponentSpec struct {
	Enabled bool          `json:"enabled,omitempty"`
	Config  PricingConfig `json:"config"`
}

// NewPricingConfig returns an empty pricing config.
func NewPricingConfig() PricingConfig {
	return PricingConfig{Sources: []PricingSourceConfig{}}
}

// NewPricingInlineSource returns an inline pricing catalog source.
func NewPricingInlineSource(catalog PricingCatalog) PricingInlineSourceConfig {
	return PricingInlineSourceConfig{Catalog: catalog}
}

// NewPricingFileSource returns a file-backed pricing catalog source.
func NewPricingFileSource(path string) PricingFileSourceConfig {
	return PricingFileSourceConfig{Path: path}
}

// NewPricingCatalog returns an inline pricing catalog with version 1.
func NewPricingCatalog(entries ...ModelPricing) PricingCatalog {
	return PricingCatalog{
		Version: 1,
		Entries: entries,
	}
}

// NewModelPricing returns a model pricing entry with catalog defaults applied.
func NewModelPricing(provider, modelID string) ModelPricing {
	return ModelPricing{
		Provider:      provider,
		ModelID:       modelID,
		Aliases:       []string{},
		Currency:      "USD",
		Unit:          PricingUnitPerToken,
		PromptCache:   NewPromptCachePricing(),
		PricingAsOf:   "",
		PricingSource: "",
	}
}

// NewTokenPricingRates returns per-token model rates.
func NewTokenPricingRates(inputPerMillion, outputPerMillion float64) TokenPricingRates {
	return TokenPricingRates{
		InputPerMillion:  inputPerMillion,
		OutputPerMillion: outputPerMillion,
	}
}

// NewPromptCachePricing returns default prompt-cache accounting settings.
func NewPromptCachePricing() PromptCachePricing {
	return PromptCachePricing{
		ReadAccounting: CacheReadAccountingIncludedInPromptTokens,
	}
}

// NewTokenRateTier returns a rate tier with the provided rates.
func NewTokenRateTier(rates TokenPricingRates) TokenRateTier {
	return TokenRateTier{Rates: rates}
}

// NewPromptTokenThresholdRateSchedule returns a full-request prompt-token threshold schedule.
func NewPromptTokenThresholdRateSchedule(tiers ...TokenRateTier) PromptTokenThresholdRateSchedule {
	if tiers == nil {
		tiers = []TokenRateTier{}
	}
	return PromptTokenThresholdRateSchedule{
		AppliesTo: "full_request",
		Tiers:     tiers,
	}
}

// NewPricingComponentSpec wraps pricing config as an enabled top-level component.
func NewPricingComponentSpec(config PricingConfig) PricingComponentSpec {
	return PricingComponentSpec{
		Enabled: true,
		Config:  config,
	}
}

// PluginComponent converts the pricing component wrapper into the shared plugin shape.
func (spec PricingComponentSpec) PluginComponent() PluginComponentSpec {
	return PluginComponentSpec{
		Kind:    PricingPluginKind,
		Enabled: spec.Enabled,
		Config:  mustConfigMap(spec.Config),
	}
}

// PricingComponent converts pricing config directly into a shared plugin component.
func PricingComponent(config PricingConfig) PluginComponentSpec {
	return NewPricingComponentSpec(config).PluginComponent()
}

// ValidatePricingConfig validates a pricing config document without activating it.
func ValidatePricingConfig(config PricingConfig) (ConfigReport, error) {
	return ValidatePluginConfig(PluginConfig{
		Version:    1,
		Components: []PluginComponentSpec{PricingComponent(config)},
	})
}

File: python/nemo_relay/pricing.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Pricing plugin configuration helpers."""

from __future__ import annotations

from dataclasses import dataclass, field, fields, is_dataclass
from typing import Literal, Protocol, TypedDict, cast

from nemo_relay import Json, JsonObject
from nemo_relay import plugin as plugin_module


class _ConfigDiagnosticRequired(TypedDict):
    level: Literal["warning", "error"]
    code: str
    message: str


class ConfigDiagnostic(_ConfigDiagnosticRequired, total=False):
    """One pricing validation diagnostic."""

    component: str
    field: str


class ConfigReport(TypedDict):
    """Validation report for pricing configuration."""

    diagnostics: list[ConfigDiagnostic]


class _SupportsToDict(Protocol):
    def to_dict(self) -> JsonObject: ...


def _normalize(value: object) -> Json:
    if hasattr(value, "to_dict"):
        return cast(_SupportsToDict, value).to_dict()
    if is_dataclass(value) and not isinstance(value, type):
        return {
            field_info.name: _normalize(field_value)
            for field_info in fields(value)
            if (field_value := getattr(value, field_info.name)) is not None
        }
    if isinstance(value, list):
        return [_normalize(item) for item in value]
    if isinstance(value, dict):
        return {cast(str, key): _normalize(val) for key, val in value.items() if val is not None}
    return cast(Json, value)


def _normalize_object(value: object) -> JsonObject:
    return cast(JsonObject, _normalize(value))


@dataclass(slots=True)
class TokenPricingRates:
    """Per-token model rates expressed per one million tokens."""

    input_per_million: float
    output_per_million: float
    cache_read_per_million: float | None = None
    cache_write_per_million: float | None = None

    def to_dict(self) -> JsonObject:
        """Serialize these rates to the canonical JSON object shape."""
        return _normalize_object(
            {
                "input_per_million": self.input_per_million,
                "output_per_million": self.output_per_million,
                "cache_read_per_million": self.cache_read_per_million,
                "cache_write_per_million": self.cache_write_per_million,
            }
        )


@dataclass(slots=True)
class PromptCachePricing:
    """Prompt-cache accounting settings for a pricing entry."""

    read_accounting: Literal["included_in_prompt_tokens", "separate"] = "included_in_prompt_tokens"

    def to_dict(self) -> JsonObject:
        """Serialize this prompt-cache config to the canonical JSON object shape."""
        return {"read_accounting": self.read_accounting}


@dataclass(slots=True)
class TokenRateTier:
    """One threshold tier in a token rate schedule."""

    rates: TokenPricingRates | JsonObject
    min_prompt_tokens: int | None = None
    max_prompt_tokens: int | None = None

    def to_dict(self) -> JsonObject:
        """Serialize this rate tier to the canonical JSON object shape."""
        return _normalize_object(
            {
                "min_prompt_tokens": self.min_prompt_tokens,
                "max_prompt_tokens": self.max_prompt_tokens,
                "rates": self.rates,
            }
        )


@dataclass(slots=True)
class PromptTokenThresholdRateSchedule:
    """Rate schedule selected by full-request prompt token thresholds."""

    tiers: list[TokenRateTier | JsonObject] = field(default_factory=list)
    applies_to: Literal["full_request"] = "full_request"

    def to_dict(self) -> JsonObject:
        """Serialize this rate schedule to the canonical JSON object shape."""
        return _normalize_object(
            {
                "type": "prompt_token_threshold",
                "applies_to": self.applies_to,
                "tiers": self.tiers,
            }
        )


@dataclass(slots=True)
class ModelPricing:
    """One model pricing catalog entry."""

    provider: str
    model_id: str
    pricing_as_of: str
    pricing_source: str
    aliases: list[str] = field(default_factory=list)
    currency: str = "USD"
    unit: Literal["per_token", "per_request", "per_second", "gpu_hour"] = "per_token"
    rates: TokenPricingRates | JsonObject | None = None
    rate_schedule: PromptTokenThresholdRateSchedule | JsonObject | None = None
    prompt_cache: PromptCachePricing | JsonObject = field(default_factory=PromptCachePricing)

    def to_dict(self) -> JsonObject:
        """Serialize this catalog entry to the canonical JSON object shape."""
        return _normalize_object(
            {
                "provider": self.provider,
                "model_id": self.model_id,
                "aliases": self.aliases,
                "currency": self.currency,
                "unit": self.unit,
                "rates": self.rates,
                "rate_schedule": self.rate_schedule,
                "prompt_cache": self.prompt_cache,
                "pricing_as_of": self.pricing_as_of,
                "pricing_source": self.pricing_source,
            }
        )


@dataclass(slots=True)
class PricingCatalog:
    """Inline pricing catalog payload."""

    entries: list[ModelPricing | JsonObject] = field(default_factory=list)
    version: int = 1

    def to_dict(self) -> JsonObject:
        """Serialize this catalog to the canonical JSON object shape."""
        return _normalize_object(
            {
                "version": self.version,
                "entries": self.entries,
            }
        )


@dataclass(slots=True)
class InlineSource:
    """Pricing source backed by an inline catalog."""

    catalog: PricingCatalog | JsonObject

    def to_dict(self) -> JsonObject:
        """Serialize this source to the canonical JSON object shape."""
        return _normalize_object(
            {
                "type": "inline",
                "catalog": self.catalog,
            }
        )


@dataclass(slots=True)
class FileSource:
    """Pricing source backed by a JSON catalog file."""

    path: str

    def to_dict(self) -> JsonObject:
        """Serialize this source to the canonical JSON object shape."""
        return {
            "type": "file",
            "path": self.path,
        }


@dataclass(slots=True)
class PricingConfig:
    """Canonical config document for the top-level pricing component."""

    sources: list[InlineSource | FileSource | JsonObject] = field(default_factory=list)

    def to_dict(self) -> JsonObject:
        """Serialize this pricing config to the canonical JSON object shape."""
        return _normalize_object({"sources": self.sources})


PRICING_PLUGIN_KIND = "pricing"


@dataclass(slots=True)
class ComponentSpec:
    """Top-level pricing component wrapper."""

    config: PricingConfig | JsonObject
    enabled: bool = True

    def to_dict(self) -> JsonObject:
        """Serialize this component to the canonical plugin shape."""
        return {
            "kind": PRICING_PLUGIN_KIND,
            "enabled": self.enabled,
            "config": _normalize_object(self.config),
        }


def validate_config(config: PricingConfig | JsonObject) -> ConfigReport:
    """Validate a pricing config document without activating it."""
    report = plugin_module.validate(
        plugin_module.PluginConfig(
            components=[ComponentSpec(config)],
        )
    )
    return cast(ConfigReport, report)


__all__ = [
    "ComponentSpec",
    "ConfigDiagnostic",
    "ConfigReport",
    "FileSource",
    "InlineSource",
    "ModelPricing",
    "PRICING_PLUGIN_KIND",
    "PromptCachePricing",
    "PromptTokenThresholdRateSchedule",
    "PricingCatalog",
    "PricingConfig",
    "TokenPricingRates",
    "TokenRateTier",
    "validate_config",
]
